service/history/ndc/hsm_state_replicator.go

```diff
@@ -0,0 +1,271 @@
+		shardContext  shard.Context
+		workflowCache wcache.Cache
+		logger        log.Logger
+	}
+)
+
+func NewHSMStateReplicator(
+	shardContext shard.Context,
+	workflowCache wcache.Cache,
+	logger log.Logger,
+) *HSMStateReplicatorImpl {
+
+	return &HSMStateReplicatorImpl{
+		shardContext:  shardContext,
+		workflowCache: workflowCache,
+		logger:        log.With(logger, tag.ComponentHSMStateReplicator),
+	}
+}
+
+func (r *HSMStateReplicatorImpl) SyncHSMState(
+	ctx context.Context,
+	request *shard.SyncHSMRequest,
+) (retError error) {
+	namespaceID := namespace.ID(request.WorkflowKey.GetNamespaceID())
+	execution := &commonpb.WorkflowExecution{
+		WorkflowId: request.WorkflowKey.GetWorkflowID(),
+		RunId:      request.WorkflowKey.GetRunID(),
+	}
+
+	lastItem, err := versionhistory.GetLastVersionHistoryItem(request.EventVersionHistory)
+	if err != nil {
+		return err
+	}
+
+	workflowContext, release, err := r.workflowCache.GetOrCreateWorkflowExecution(
+		ctx,
+		r.shardContext,
+		namespaceID,
+		execution,
+		workflow.LockPriorityHigh,
+	)
+	if err != nil {
+		return err
+	}
+	defer func() { release(retError) }()
+
+	mutableState, err := workflowContext.LoadMutableState(ctx, r.shardContext)
+	if err != nil {
+		if _, isNotFound := err.(*serviceerror.NotFound); isNotFound {
+			return serviceerrors.NewRetryReplication(
+				"sync HSM state encountered workflow not found",
+				namespaceID.String(),
+				execution.GetWorkflowId(),
+				execution.GetRunId(),
+				common.EmptyEventID,
+				common.EmptyVersion,
+				lastItem.EventId+1, // underlying resend logic is exclusive-exclusive
+				lastItem.Version,
+			)
+		}
+		return err
+	}
+
+	synced, err := r.syncHSMNode(mutableState, request)
+	if err != nil || !synced {
+		return err
+	}
+
+	state, _ := mutableState.GetWorkflowStateStatus()
+	if state == enumsspb.WORKFLOW_EXECUTION_STATE_COMPLETED {
+		return workflowContext.SubmitClosedWorkflowSnapshot(
+			ctx,
+			r.shardContext,
+			workflow.TransactionPolicyPassive,
+		)
+	}
+
+	updateMode := persistence.UpdateWorkflowModeUpdateCurrent
+	if state == enumsspb.WORKFLOW_EXECUTION_STATE_ZOMBIE {
+		updateMode = persistence.UpdateWorkflowModeBypassCurrent
+	}
+
+	return workflowContext.UpdateWorkflowExecutionWithNew(
+		ctx,
+		r.shardContext,
+		updateMode,
+		nil, // no new workflow
+		nil, // no new workflow
+		workflow.TransactionPolicyPassive,
+		nil,
+	)
+}
+
+func (r *HSMStateReplicatorImpl) syncHSMNode(
+	mutableState workflow.MutableState,
+	request *shard.SyncHSMRequest,
+) (bool, error) {
+
+	shouldSync, err := r.compareVersionHistory(mutableState, request.EventVersionHistory)
+	if err != nil || !shouldSync {
+		return shouldSync, err
+	}
+
+	currentHSM := mutableState.HSM()
+
+	// we don't care about the root here which is the entire mutable state
+	incomingHSM, err := hsm.NewRoot(
+		r.shardContext.StateMachineRegistry(),
+		workflow.StateMachineType.ID,
+		mutableState,
+		request.StateMachineNode.Children,
```

Comment:
You could have just synced the children instead of the containing node (which is otherwise empty) but it's not critical.